Feature simplified ribbon

nishy2000 opened the pull request:
I am proceeding with implementation based on PR #707 for issue #695.
This implement resolves #695 (a part of full feature) 

Note: In the showcase program, the Simplified Ribbon can be found under Tests --> Windows --> "Open Simplified Ribbon-Window"

First of all, I am currently implementing the following items.
- Add CanUseSimplified property in the ribbon control. This property is to control adding GUI for switching between classic and simplified ribbons.
- Add IsSimplified property to the Ribbon control
- Add a read-only IsSimplified property for switching  styles to controls that support simplified mode
- Add a MediumIcon property to controls that support simplified mode
- Add ControlTemplate for simplified mode to controls that support simplified mode
- Add a SimplifiedSizeDefinition property to controls that support simplified mode as the SizeDefinition property for Simplified mode.
- Add a SimplifiedStateDefinition property to RibbonGroupBox control as the StateDefinition property for Simplified mode. The default value of StateDefinition is "Large,Middle,Small,Collapased", but the default value of SimplifiedStateDefinition is "Large,Middle,Collapased".

Overflow button is not supported in this PR. I will  create another PR to support the overflow button.

The basic idea on Simplified mode is that the child controls of the RibbonGroupBox show the elements as follows:
- If RibbonControl.Size is Large, show MediumIcon (24x24) and Header
- If RibbonControl.Size is Middle, show MediumIcon (24x24)
- if RibbonControl.Size is Small, show Icon (16x16)

batzen commented:
@nishy2000 I added a new `IconPresenter` control.
Maybe that makes your work easier. ;-)

nishy2000 commented:
@batzen , Thank you, I think it will simplify the control style. I will try to modify to use the `IconPresenter` control.

batzen commented:
@nishy2000 Are you still working on this or should i start reviewing your changes?

nishy2000 commented:
@batzen 
I thought I was done with the implementation.
However, today I realized that there is an omission in the implementation: I didn't add the GUI for switching between the classic ribbon and the simplified ribbon. I think the switching GUI is a must-have feature. Can you please wait a bit?

batzen commented:
@nishy2000 Of course. Didn't want to pressure you. Take your time and just ping me if you are ready.

Splitwirez commented:
> @batzen
> I thought I was done with the implementation.
> However, today I realized that there is an omission in the implementation: I didn't add the GUI for switching between the classic ribbon and the simplified ribbon. I think the switching GUI is a must-have feature. Can you please wait a bit?

It might be, but...I'd recommend thinking twice before assuming that both GUIs are adequately-suited to every use-case.

nishy2000 commented:
> @nishy2000 It might be, but...I'd recommend thinking twice before assuming that both GUIs are adequately-suited to every use-case.

@Splitwirez 
That's right. I'm thinking of the following.
Add a new property to the ribbon control to enable the new switching UI(context menu and switching button). If the new property of the ribbon control is not set in XAML, the behavior will be the same as before.

nishy2000 commented:
@batzen
I have finished. Please start review.

batzen commented:
@nishy2000 Will try to have a look during the weekend.

batzen commented:
@nishy2000 I started my review during the weekend but didn't have time to finish it. Will try to get it done during the week. Sorry for the delay.

nishy2000 commented on Fluent.Ribbon.Showcase/Fluent.Ribbon.Showcase.csproj:
```diff
@@ -40,4 +43,9 @@
+    <Compile Update="SimplifiedRibbonWindow.xaml.cs">
+      <SubType>Code</SubType>
+    </Compile>
+  </ItemGroup>
```
Thank you.
I have fixed at commit https://github.com/fluentribbon/Fluent.Ribbon/pull/924/commits/91777cc9c00d3a4dfaa2426dcde4611aecd73b0b

batzen commented:
Sorry again for the delay.
Most of the changes look good to me.

There are a couple of things i don't really like.
One part is the "iconImageBorder" you introduced in some controls.
I don't think we need that and if people want a border when not providing an icon they could provide the border as an icon.

The other part are the separate templates.
In don't really get why we need separate templates for every control.
The ComboBox templates, for example, don't really differ in content and just add more maintenance cost (and risk for bugs).
I guess the reason for you to create a separate template, in case of the ComboBox, was to get the mouse over border showing, but that's not what happens in the simplified ribbon in Word Online (which is the only Word version that i could find that has a simplified ribbon).

Which were the reasons to create separate templates for all the other controls?
Most things should be solvable through attached properties.

And thanks for all your great work and your patience. 
I currently struggle to find motivation to work on OSS, that's the reason for the long delays.

nishy2000 commented:
@batzen 
Thanks for your review.

**About iconImageBorder:**
I will look into your suggestion about IconBorder (border as icon).

**About separate templates:**
If we talk only about ComboBox, I added Border not for mouseover handling, but for Padding handling.
To make it easier to compare the differences in the ComboBox templates, I added the commit 58f65f9 that matches the order of the properties. Please check it.

Overall I separated the templates for a several reasons.

- Compatibility with existing behaviors
- Avoidance of merge conflicts
- Increasing complexity of triggers

The first is to avoid affecting the existing behavior (classic ribbon). I wanted to avoid adding functionality to the Simple Ribbon that would affect the Classic Ribbon's functionality.

The second is to avoid conflicts. During the implementation of the Simple Ribbon, bug fixes and feature improvements to Fluent.Ribbon continued. If a single template is used to deal with the problem, it is difficult to understand the differences when merging, and conflicts are more likely to occur.

Third, the complexity of triggers is increasing. If we use a single template, IsSimplified will be added to the trigger condition, making it a multi-condition instead of a single condition. I thought this might be a complication.

However, if you think the above reason is not a big deal, I can integrate the template.

In fact,  for example, I was concerned about the existence of duplicate code in the ComboBox template, since the definition of the Popup control is exactly the same. Therefore, I thought of a countermeasure. However, I couldn't even change the Popup definition to a style definition because it is referenced in Template.Triggers.
Do you have any good ideas on this point?

batzen commented:
I have to admit that the triggers are really a mess already and sometimes way too complex.
But maintaining triggers is still easier than maintaining complete different templates.
I didn't compare all the templates, but in case of the ComboBox i don't think we need a separate template.

What do you think about this suggestion:
I will merge your changes, so you don't have to constantly merge changes from develop and resolve conflicts.
That way people also get the chance to test the simplified ribbon.
I will try to check all templates and remove separated ones where we don't need them because the differences are so small.
Some things might be solvable by introducing new attached properties. Maybe it's really easier to have separated templates in some places, i am still not sure about that.

And i should introduce something like XAML-Styler to have all XAML formatted in the same way. That way comparing everything should get a lot easier. That should be the first thing for me to do after your PR is merged.

@nishy2000 Just tell me when i can merge. And sorry again for the long delay.

nishy2000 commented:
@batzen 
I have finished all. Please merge PR. I'm looking forward to it.

batzen commented:
I will resolve the conflicts tomorrow and merge.
